src/d3d9/shaders/d3d9_fixed_function_frag.frag

```diff
@@ -0,0 +1,667 @@
+        previousStageColorOp == D3DTOP_BUMPENVMAP
+        || previousStageColorOp == D3DTOP_BUMPENVMAPLUMINANCE)) {
+        texcoord = calculateBumpmapCoords(stage, texcoord, previousStageTextureVal);
+    }
+
+    vec4 texVal;
+    uint textureType = D3DRTYPE_TEXTURE + specUint(SpecSamplerType, 2u * stage, 2u);
+    switch (textureType) {
+        case D3DRTYPE_TEXTURE:
+            if (specBool(SpecSamplerDepthMode, stage))
+                texVal = texture(sampler2DShadow(t2d[stage], sampler_heap[loadSamplerHeapIndex(stage)]), adjustDref(texcoord, 2, stage).xyz).xxxx;
+            else
+                texVal = texture(sampler2D(t2d[stage], sampler_heap[loadSamplerHeapIndex(stage)]), texcoord.xy);
+            break;
+        case D3DRTYPE_CUBETEXTURE:
+            if (specBool(SpecSamplerDepthMode, stage))
+                texVal = texture(samplerCubeShadow(tcube[stage], sampler_heap[loadSamplerHeapIndex(stage)]), adjustDref(texcoord, 3, stage)).xxxx;
+            else
+                texVal = texture(samplerCube(tcube[stage], sampler_heap[loadSamplerHeapIndex(stage)]), texcoord.xyz);
+            break;
+        case D3DRTYPE_VOLUMETEXTURE:
+            texVal = texture(sampler3D(t3d[stage], sampler_heap[loadSamplerHeapIndex(stage)]), texcoord.xyz);
+            break;
+    }
+
+    if (stage != 0 && previousStageColorOp == D3DTOP_BUMPENVMAPLUMINANCE) {
+        float lScale = sharedData.Stages[stage - 1].BumpEnvLScale;
+        float lOffset = sharedData.Stages[stage - 1].BumpEnvLOffset;
+        float scale = texVal.z;
+        scale *= lScale;
+        scale += lOffset;
+        scale = clamp(scale, 0.0, 1.0);
+        texVal *= scale;
+    }
+
+    return texVal;
+}
+
+
+vec4 readArgValue(uint stage, uint arg, vec4 current, vec4 temp, vec4 textureVal) {
+    vec4 reg = vec4(1.0);
+    switch (arg & D3DTA_SELECTMASK) {
+        case D3DTA_CONSTANT:
+            reg = vec4(
+                sharedData.Stages[stage].Constant[0],
+                sharedData.Stages[stage].Constant[1],
+                sharedData.Stages[stage].Constant[2],
+                sharedData.Stages[stage].Constant[3]
+            );
+            break;
+        case D3DTA_CURRENT:
+            reg = current;
+            break;
+        case D3DTA_DIFFUSE:
+            reg = in_Color0;
+            break;
+        case D3DTA_SPECULAR:
+            reg = in_Color1;
+            break;
+        case D3DTA_TEMP:
+            reg = temp;
+            break;
+        case D3DTA_TEXTURE:
+            reg = textureVal;
+            break;
+        case D3DTA_TFACTOR:
+            reg = data.textureFactor;
+            break;
+    }
+
+    // reg = 1 - reg
+    if ((arg & D3DTA_COMPLEMENT) != 0)
+        reg = vec4(1.0) - reg;
+
+    // reg = reg.wwww
+    if ((arg & D3DTA_ALPHAREPLICATE) != 0)
+        reg = reg.aaaa;
+
+    return reg;
+}
+
+vec4[TextureArgCount] readArgValues(uint stage, uint args[TextureArgCount], vec4 current, vec4 temp, vec4 textureVal) {
```

Comment:
I'm still rather allergic to this being an array, there doesn't seem to be any actual dynamic indexing going on for these anyway so can this just be changed into a struct?